XDMA/linux-kernel/xdma/xdma_netdev.c

```diff
@@ -281,55 +285,63 @@ static void do_tx_work(struct work_struct *work, u16 tstamp_id) {
         }
 
         if (!priv->tx_work_skb[tstamp_id]) {
-                clear_bit_unlock(tstamp_id, &priv->state);
-                return;
+                goto return_error;
         }
 
         if (now < priv->tx_work_start_after[tstamp_id]) {
-                schedule_work(&priv->tx_work[tstamp_id]);
-                return;
+                goto retry;
         }
         /*
          * Read TX timestamp several times because
          * 1. Reading and writing TX timestamp register are not atomic
          * 2. The work thread might try to read TX timestamp before the register gets updated
          */
         tx_tstamp = alinx_read_tx_timestamp(priv->pdev, tstamp_id);
-        if (tx_tstamp == priv->last_tx_tstamp[tstamp_id] && priv->tstamp_retry[tstamp_id] < TX_TSTAMP_MAX_RETRY) {
+        if (tx_tstamp == priv->last_tx_tstamp[tstamp_id]) {
                 if (alinx_get_sys_clock(priv->xdev->pdev) < priv->tx_work_wait_until[tstamp_id]) {
                         /* The packet might have not been sent yet */
-                        schedule_work(&priv->tx_work[tstamp_id]);
-                        return;
+                        goto retry;
                 }
                 /*
                  * Tx timestamp is not updated. Try again.
                  * Waiting for it to be updated forever is not desirable,
                  * so limit the number of retries
                  */
-                priv->tstamp_retry[tstamp_id]++;
-                if (priv->tstamp_retry[tstamp_id] >= TX_TSTAMP_MAX_RETRY) {
+                if (++priv->tstamp_retry[tstamp_id] >= TX_TSTAMP_MAX_RETRY) {
                         /* TODO: track the number of skipped packets for ethtool stats */
-                        priv->tstamp_retry[tstamp_id] = 0;
-                        clear_bit_unlock(tstamp_id, &priv->state);
-                        return;
+                        pr_warn("Failed to get timestamp: timestamp is not getting updated, " \
+                                "the packet might have been dropped\n");
+                        goto return_error;
                 }
-                schedule_work(&priv->tx_work[tstamp_id]);
-                return;
+                goto retry;
+        } else if (now - tx_tstamp > TX_TSTAMP_UPDATE_THRESHOLD) {
+                /* Tx timestamp is only partially updated */
+                if (++priv->tstamp_retry[tstamp_id] >= TX_TSTAMP_MAX_RETRY) {
```

Comment:
```suggestion
                if (++(priv->tstamp_retry[tstamp_id]) >= TX_TSTAMP_MAX_RETRY) {
```

애매한 연산자 우선순위에 대해선 괄호를 씌우는 게 좋죠. 위쪽 310L도 마찬가지입니다.

Reply:
수정했습니다.